npm/pod.go

```diff
@@ -32,222 +94,374 @@
 		oldPodObj.Status.PodIP == newPodObj.Status.PodIP &&
 		newPodObj.ObjectMeta.DeletionTimestamp == nil &&
 		newPodObj.ObjectMeta.DeletionGracePeriodSeconds == nil
-	isInvalidUpdate = isInvalidUpdate && reflect.DeepEqual(oldPodObj.ObjectMeta.Labels, newPodObj.ObjectMeta.Labels)
+	isInvalidUpdate = isInvalidUpdate &&
+		reflect.DeepEqual(oldPodObj.ObjectMeta.Labels, newPodObj.ObjectMeta.Labels) &&
+		reflect.DeepEqual(oldPodObj.Status.PodIPs, newPodObj.Status.PodIPs) &&
+		reflect.DeepEqual(getContainerPortList(oldPodObj), getContainerPortList(newPodObj))
 
 	return
 }
 
+func getContainerPortList(podObj *corev1.Pod) []v1.ContainerPort {
+	portList := []v1.ContainerPort{}
+	for _, container := range podObj.Spec.Containers {
+		portList = append(portList, container.Ports...)
+	}
+	return portList
+}
+
+// appendNamedPortIpsets helps with adding or deleting Pod namedPort IPsets
+func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPort, podUID string, podIP string, delete bool) error {
+
+	for _, port := range portList {
+		if port.Name == "" {
+			continue
+		}
+
+		protocol := ""
+
+		switch port.Protocol {
+		case v1.ProtocolUDP:
+			protocol = util.IpsetUDPFlag
+		case v1.ProtocolSCTP:
+			protocol = util.IpsetSCTPFlag
+		case v1.ProtocolTCP:
+			protocol = util.IpsetTCPFlag
+		}
+
+		namedPortname := util.NamedPortIPSetPrefix + port.Name
+
+		if delete {
+			// Delete the pod's named ports from its ipset.
+			ipsMgr.DeleteFromSet(
+				namedPortname,
+				fmt.Sprintf("%s,%s%d", podIP, protocol, port.ContainerPort),
+				podUID,
+			)
+			continue
+		}
+		// Add the pod's named ports to its ipset.
+		ipsMgr.AddToSet(
+			namedPortname,
+			fmt.Sprintf("%s,%s%d", podIP, protocol, port.ContainerPort),
+			util.IpsetIPPortHashFlag,
+			podUID,
+		)
+
+	}
+
+	return nil
+}
+
 // AddPod handles adding pod ip to its label's ipset.
 func (npMgr *NetworkPolicyManager) AddPod(podObj *corev1.Pod) error {
 	if !isValidPod(podObj) {
 		return nil
 	}
 
+	// Ignore adding the HostNetwork pod to any ipsets.
+	if isHostNetworkPod(podObj) {
+		log.Logf("HostNetwork POD IGNORED: [%s/%s/%s/%+v%s]", podObj.GetObjectMeta().GetUID(), podObj.Namespace, podObj.Name, podObj.Labels, podObj.Status.PodIP)
+		return nil
+	}
+
+	if podObj.Status.Phase == v1.PodSucceeded || podObj.Status.Phase == v1.PodFailed {
```

Comment:
Can you add a comment for why this is blocked?

Reply:
Done